source/common/router/delegating_route_impl.h

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#include "envoy/router/router.h"
+
+namespace Envoy {
+namespace Router {
+
+/**
+ * Implementation of DelegatingRoute that delegates all method calls to the RouteConstSharedPtr
+ * base route it wraps around.
+ *
+ * Intended to be used as a route mutability mechanism, where a filter can create a derived class of
+ * DelegatingRoute and override specific methods (e.g. routeEntry) while preserving the rest of the
+ * properties/behavior of the base route.
+ */
+class DelegatingRoute : public Router::Route {
+public:
+  DelegatingRoute(Router::RouteConstSharedPtr route) : base_route_(route) {}
+
+  const Router::DirectResponseEntry* directResponseEntry() const override;
+  const Router::RouteEntry* routeEntry() const override;
+  const Router::Decorator* decorator() const override;
+  const Router::RouteTracing* tracingConfig() const override;
+  const Router::RouteSpecificFilterConfig* perFilterConfig(const std::string&) const override;
+
+private:
+  Router::RouteConstSharedPtr base_route_;
+};
+
+class DelegatingRouteEntry : public Router::RouteEntry {
+public:
+  DelegatingRouteEntry(const Router::RouteEntry* route_entry) : base_route_entry_(route_entry) {}
```

Comment:
The code in the .cc file assumes that this is nullptr, so passing `const Router::RouteEntry&` seems better.

We should probably document the lifetime guarantees and perhaps talk about why its fine for this class to take the delegated object as a non-owning ref while the `DelegatingRoute` takes a shared_ptr

Reply:
1. Hmm sorry what do you mean by "assumes that this is nullptr"? I had it using `const Router::RouteEntry*` to align with `const Router::RouteEntry* routeEntry() const override;` from the `DelegatingRoute` class.

2. Agree that we should document lifetime guarantees and ownership choices made. Let me think about this for a bit. What did you have in mind? I chose `DelegatingRoute` to take a shared_ptr to align with existing route functionality/methods such as [route() returning a RouteConstSharedPtr ](https://github.com/envoyproxy/envoy/blob/v1.17.1/include/envoy/http/filter.h#L197)

Reply:
The code in the cc file dereferences the pointer without any nulllpointer checks, so if you passed nullptr here you would segfault. Either we can never pass nullptr to this class (in which case taking a const& is recommended) or we have to add null checks to all the functions. I prefer the const& approach since it seems nonsensical to delegate to a null route (just pass the null route)

For lifetime I think its fine to copy around a shared ptr if you already have one, though if you want to pass a raw pointer/ref around we need to document the expected lifetime of it, since nothing in code prevents that from being deallocated somewhere else while we're still holding onto the pointer/ref. This could be as easy as saying "X must outlive Y" in comments, though we'll want to make sure that's actually the case

Reply:
Ah, I see. Yeah I agree `const Router::RouteEntry&` is the right decision here, since it is nonsensical to delegate to a null route. 

Re: lifetime, got it. We could document something like `@param route_entry must outlive the lifetime of this DelegatingRouteEntry class`. I don't think it'd get deallocated somewhere else, but is there a way to really make sure? Would it be smarter to play it safe and have it be a ConstSharedPtr like the `DelegatingRoute` constructor? But then we'd run into the original issue of needing to add null checks to all the functions.

Reply:
Actually, after chatting with @theevocater about this, we think it could be a good idea to change the "base" that `DelegatingRouteEntry` uses for delegation from a `const Router::RouteEntry&` to a `Router::RouteConstSharedPtr` (done in commit https://github.com/envoyproxy/envoy/pull/15266/commits/27c42ced9e4bd690105845de543f61155e7f110d) 

The ctor is cleaner, we don't need to worry about the raw pointer/ref lifetime of RouteEntry, and `RouteConstSharedPtr` was meant to be shared. What do you think?